File: src/test/java/com/mmust/evaluators/PostfixMathExpressionEvaluatorTest.java
package com.mmust.evaluators;

import com.mmust.evaluators.impl.StackPostfixMathExpressionEvaluator;
import com.mmust.token.NumToken;
import com.mmust.token.Token;
import org.junit.Before;
import org.junit.Test;

import java.util.List;

import static com.mmust.token.Operators.*;
import static org.assertj.core.api.Assertions.assertThat;

public class PostfixMathExpressionEvaluatorTest {

	private PostfixEvaluator evaluator;

	@Before
	public void init() {
		evaluator = new StackPostfixMathExpressionEvaluator();
	}

	@Test
	public void testAddition() {
		Iterable<Token> simplePostfixExpr = List.of(
				new NumToken("2"),
				new NumToken("3"),
				PLUS
		);

		double actualResult = evaluator.eval(simplePostfixExpr);
		double expectedResult = 5.0;

		assertThat(actualResult)
				.isEqualTo(expectedResult);
	}

	@Test
	public void testSubtract() {
		Iterable<Token> simplePostfixExpr = List.of(
				new NumToken("5"),
				new NumToken("2"),
				MINUS
		);

		double actualResult = evaluator.eval(simplePostfixExpr);
		double expectedResult = 3.0;

		assertThat(actualResult)
				.isEqualTo(expectedResult);
	}

	@Test
	public void testMultiply() {
		Iterable<Token> simplePostfixExpr = List.of(
				new NumToken("3"),
				new NumToken("4"),
				MULTIPLY
		);

		double actualResult = evaluator.eval(simplePostfixExpr);
		double expectedResult = 12.0;

		assertThat(actualResult)
				.isEqualTo(expectedResult);
	}

	@Test
	public void testDivide() {
		Iterable<Token> simplePostfixExpr = List.of(
				new NumToken("8"),
				new NumToken("4"),
				DIV
		);

		double actualResult = evaluator.eval(simplePostfixExpr);
		double expectedResult = 2.0;

		assertThat(actualResult)
				.isEqualTo(expectedResult);
	}

	@Test
	public void testFewOps() {
		Iterable<Token> simplePostfixExpr = List.of(
				new NumToken("3"),
				new NumToken("4"),
				new NumToken("5"),
				MULTIPLY,
				MINUS
		);

		double actualResult = evaluator.eval(simplePostfixExpr);
		double expectedResult = -17.0;

		assertThat(actualResult)
				.isEqualTo(expectedResult);
	}
}

File: src/test/kotlin/com/mmust/evaluators/PostfixMathExpressionEvaluatorTest.kt
package com.mmust.evaluators

import com.mmust.evaluators.impl.StackPostfixMathExpressionEvaluator
import com.mmust.token.NumToken
import com.mmust.token.Operators.*
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

class PostfixMathExpressionEvaluatorTest {

    private val evaluator: PostfixEvaluator = StackPostfixMathExpressionEvaluator()

    @Test
    @DisplayName("+")
    fun testAddition() {
        val simplePostfixExpr = listOf(
                NumToken("2"),
                NumToken("3"),
                PLUS
        )

        val actualResult = evaluator.eval(simplePostfixExpr)
        val expectedResult = 5.0

        assertThat(actualResult)
                .isEqualTo(expectedResult)
    }

    @Test
    @DisplayName("-")
    fun testSubtract() {
        val simplePostfixExpr = listOf(
                NumToken("5"),
                NumToken("2"),
                MINUS
        )

        val actualResult = evaluator.eval(simplePostfixExpr)
        val expectedResult = 3.0

        assertThat(actualResult)
                .isEqualTo(expectedResult)
    }

    @Test
    @DisplayName("*")
    fun testMultiply() {
        val simplePostfixExpr = listOf(
                NumToken("3"),
                NumToken("4"),
                MULTIPLY
        )

        val actualResult = evaluator.eval(simplePostfixExpr)
        val expectedResult = 12.0

        assertThat(actualResult)
                .isEqualTo(expectedResult)
    }

    @Test
    @DisplayName("/")
    fun testDivide() {
        val simplePostfixExpr = listOf(
                NumToken("8"),
                NumToken("4"),
                DIV
        )

        val actualResult = evaluator.eval(simplePostfixExpr)
        val expectedResult = 2.0

        assertThat(actualResult)
                .isEqualTo(expectedResult)
    }

    @Test
    @DisplayName("Few operations")
    fun testFewOps() {
        val simplePostfixExpr = listOf(
                NumToken("3"),
                NumToken("4"),
                NumToken("5"),
                MULTIPLY,
                MINUS
        )

        val actualResult = evaluator.eval(simplePostfixExpr)
        val expectedResult = -17.0

        assertThat(actualResult)
                .isEqualTo(expectedResult)
    }
}